### PR TITLE
LEST-92 - Make docs a TypeScript project

### PR DESCRIPTION
**Pull request by yogwoggf**
## Ticket

Resolves [LEST-92](https://taservers.atlassian.net/browse/LEST-92)

## Changes

- Made docs a TS project.
- Moved all docs to `src/`
- Changed all `index.js` to TypeScript files which verify the shape of the given JSON docs at compile-time.
- Made the site use the docs interfaces

## Impact

This allows for the docs package to *finally* play nice with other packages. Good to note however that there's now a DX overhead where a developer will have to add two lines to the `index.ts` of the section of the docs that they're editing now. (see any of the index.ts files in docs)

I did figure out that I could bypass that by using `require`, but it didn't feel good so I just kept it to the actual JSON imports section and then the export section.

## Testing

You can fire up the Docusaurus site and check for yourself that everything is working. You can also try importing anything from the docs.

## Helpful Links

[API Docs](https://taservers.github.io/lest/docs/api/expect)

[Project Board](https://taservers.atlassian.net/jira/software/c/projects/LEST/boards/4)

[Discord](https://discord.tasevers.com)

[LEST-92]: https://taservers.atlassian.net/browse/LEST-92?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by yogwoggf** [comment]
@Derpius Do you want me to create the CI? You seem to take over the CI in each ticket.

File: packages/docs/package.json [hunk 1]
```diff
@@ -1,5 +1,11 @@
 {
   "name": "@lest/docs",
   "private": true,
-  "description": "API documentation in JSON form."
+  "description": "API documentation in JSON form.",
+  "main": "dist/src/index.js",
```
Comment (Derpius) [inline packages/docs/package.json]: You have a misconfigured tsconfig if src/index is not emitting to dist/index

File: packages/docs/package.json [hunk 2]
```diff
@@ -1,5 +1,11 @@
...
+  "type": "commonjs",
+  "scripts": {
+    "build": "tsc",
+    "preinstall": "npm run build"
```
Comment (Derpius) [inline packages/docs/package.json]: Why was this added? You shouldn’t need to build dependencies manually

File: packages/docs/src/interfaces/class.ts [hunk 3]
```diff
@@ -1,6 +1,5 @@
 import { Property } from "./property";
 import { Function } from "./function";
-
```
Comment (Derpius) [inline packages/docs/src/interfaces/class.ts]: Why was this newline removed?

File: packages/docs/src/interfaces/function.ts [hunk 4]
```diff
@@ -1,5 +1,4 @@
-import { Property } from "@site/src/doc-types/property";
-
```
Comment (Derpius) [inline packages/docs/src/interfaces/function.ts:2]: Similar to above, why was this newline removed?

File: packages/docs/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
+  "compilerOptions": {
+    "target": "ES2020",
+    "composite": true,
```
Comment (Derpius) [inline packages/docs/tsconfig.json]: Is composite correct here? I don’t think we have a top level tsconfig at the moment

File: packages/docs/src/matchers/symmetric/index.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import * as _toHaveReturnedTimes from "./toHaveReturnedTimes.json";
+import * as _toHaveReturnedWith from "./toHaveReturnedWith.json";
+import * as _toHaveLastReturnedWith from "./toHaveLastReturnedWith.json";
+import * as _toHaveNthReturnedWith from "./toHaveNthReturnedWith.json";
+import * as _toHaveLength from "./toHaveLength.json";
+import * as _toMatch from "./toMatch.json";
+import * as _toMatchObject from "./toMatchObject.json";
+import * as _toThrow from "./toThrow.json";
+
+export const toBe: Function = _toBe;
```
Comment (Derpius) [inline packages/docs/src/matchers/symmetric/index.ts:34]: Is there no way to retype the export inline with the above?

i.e. something equivalent to:
```ts
export { default as blah } from "./blah.json" as Function;
```

File: packages/docs/src/matchers/symmetric/index.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import { Function } from "../../interfaces";
+
+import * as _toBe from "./toBe.json";
```
Comment (Derpius) [inline packages/docs/src/matchers/symmetric/index.ts]: We can just do `import blah from "blah.json";`

File: packages/site/package.json [hunk 8]
```diff
@@ -16,7 +16,7 @@
   "dependencies": {
     "@docusaurus/core": "2.4.0",
     "@docusaurus/preset-classic": "2.4.0",
-    "@lest/docs": "*",
+    "@lest/docs": "file:../docs",
```
Comment (Derpius) [inline packages/site/package.json]: Why was this needed?

File: packages/site/src/components/FunctionRenderer/FunctionRenderer.tsx [hunk 9]
```diff
@@ -18,16 +18,18 @@ export const FunctionRenderer: React.FC<FunctionRendererProps> = ({
...
-				{returns.length > 0 && `: ${renderReturnSignature(returns)}`}
-			</code>
-		</Heading>
-		<Aliases aliases={aliases} />
-		<Description text={description} />
-		{children}
-	</section>
-);
+}) => {
+	return (
```
Comment (Derpius) [inline packages/site/src/components/FunctionRenderer/FunctionRenderer.tsx]: Why did you make this change?

File: packages/docs/package.json [hunk 10]
```diff
@@ -1,5 +1,11 @@
 {
   "name": "@lest/docs",
   "private": true,
-  "description": "API documentation in JSON form."
+  "description": "API documentation in JSON form.",
+  "main": "dist/src/index.js",
+  "type": "commonjs",
+  "scripts": {
```
Comment (Derpius) [inline packages/docs/package.json:7]: Could you also add a `format` script which runs prettier on the JSON and TS?

In the future we probably also want to assert that all json docs are exported from the top level, but that would need Jest set up and a test written to detect all JSON docs.

File: packages/docs/package.json (hunk 1 above)
Comment (yogwoggf) [reply packages/docs/package.json]: Ok!

File: packages/docs/tsconfig.json (hunk 5 above)
Comment (yogwoggf) [reply packages/docs/tsconfig.json]: Well, it's used for when you have a primarily TypeScript monorepo structure. I think it's safe to stay.

Comment (Derpius) [reply packages/docs/tsconfig.json]: https://www.typescriptlang.org/tsconfig#composite

Not sure if it's actually doing anything here though

File: packages/docs/src/matchers/symmetric/index.ts (hunk 6 above)
Comment (yogwoggf) [reply packages/docs/src/matchers/symmetric/index.ts:34]: Not that I know of.

1. require does not resolve the types at compile, so we lose our strict typechecking of the docs.
2. `export { default as blah from "./blah" }` is unable to be typed.

File: packages/docs/src/matchers/symmetric/index.ts (hunk 7 above)
Comment (yogwoggf) [reply packages/docs/src/matchers/symmetric/index.ts]: Tried this, and for the Docusaurus site it returns undefined for each import. Apparently this is an issue with webpack.

Comment (Derpius) [reply packages/docs/src/matchers/symmetric/index.ts]: Assign me to the PR once you've pushed markups for everything else and I'll take a look

File: packages/site/src/components/FunctionRenderer/FunctionRenderer.tsx (hunk 9 above)
Comment (yogwoggf) [reply packages/site/src/components/FunctionRenderer/FunctionRenderer.tsx]: Debug code

File: packages/docs/tsconfig.json (hunk 5 above)
Comment (yogwoggf) [reply packages/docs/tsconfig.json]: In this scenario we're enforcing that:
- declarations must be generated for each build (so ts packages including any given package will receive the types)
- all files must be in the `include` array
- and that the rootDir defaults to the child package (top-level tsconfig)

This is all done by `composite`

Comment (yogwoggf) [reply packages/docs/tsconfig.json]: And... further research shows that declarations aren't generated despite it being on the handbook..

**Comment by yogwoggf** [comment]
Ok, here are the changes from the past few commits:
- Added top-level tsconfig that each primary TS project inherits from. (docusaurus inherits its own tsconfig, and it's browser so it's not using it as of now)
- Resolved markups except for the ones that need discussion

File: packages/lest/tsconfig.json [hunk 11]
```diff
@@ -1,13 +1,8 @@
...
-    "target": "es5",
-    "sourceMap": true,
-    "declaration": true,
     "strict": true,
-    "esModuleInterop": true,
-    "resolveJsonModule": true,
     "outDir": "dist"
   },
-  "include": ["src/**/*.ts", "tests/**/*.ts"]
+  "include": ["src/**/*.ts", "src/**/*.json", "tests/**/*.ts"]
```
Comment (Derpius) [inline packages/lest/tsconfig.json]: Why was this added to the lest project?

File: packages/lest/tsconfig.json [hunk 12]
```diff
@@ -1,13 +1,8 @@
 {
+  "extends": "../../tsconfig.json",
   "compilerOptions": {
-    "module": "commonjs",
-    "target": "es5",
-    "sourceMap": true,
-    "declaration": true,
     "strict": true,
```
Comment (Derpius) [inline packages/lest/tsconfig.json]: Shouldn't `strict` be enabled in the root config?

File: packages/lest/tsconfig.build.json [hunk 13]
```diff
@@ -1,4 +1,4 @@
 {
   "extends": "./tsconfig.json",
-  "include": ["src/**/*.ts"]
+  "include": ["src/**/*.ts", "src/**/*.json"]
```
Comment (Derpius) [inline packages/lest/tsconfig.build.json]: Why was this added to the lest project?

File: packages/site/src/components/object-renderer/object.tsx [hunk 14]
```diff
@@ -1,8 +1,8 @@
 import React, { createContext, useContext } from "react";
 import Heading from "@theme/Heading";
 import Description from "../Description";
-import { Class } from "../../doc-types/class";
-import { isTableProperty, Property } from "../../doc-types";
+import { Class } from "@lest/docs";
+import { isTableProperty, Property } from "@lest/docs";
```
Comment (Derpius) [inline packages/site/src/components/object-renderer/object.tsx]: nit: We can combine these import statements

File: tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,13 @@
+{
+  "compilerOptions": {
+    "target": "ES2020",
+    "composite": true,
```
Comment (Derpius) [inline tsconfig.json]: I've had a look at some of my typescript monorepos, this is not required. Do you have a functional reason for including this that affects the project? Generally when you stray from the `tsconfig.json` defaults it causes more issues...

File: packages/docs/package.json [hunk 16]
```diff
@@ -1,5 +1,12 @@
 {
   "name": "@lest/docs",
   "private": true,
-  "description": "API documentation in JSON form."
+  "description": "API documentation in JSON form.",
+  "main": "dist/index.js",
+  "type": "commonjs",
+  "scripts": {
+    "build": "tsc",
+    "preinstall": "npm run build",
```
Comment (Derpius) [inline packages/docs/package.json]: Is this still required with the top level tsconfig? 

I have not had to do this in any TS monorepo...

File: tsconfig.json [hunk 17]
```diff
@@ -0,0 +1,13 @@
+{
+  "compilerOptions": {
```
Comment (Derpius) [inline tsconfig.json:2]: I think we need a separate `tsconfig.json` and `tsconfig.build.json` at the top level, with the top level `tsconfig.json` extending `tsconfig.build.json` and defining the `paths` property:

```json
    "paths": {
      "@lest/*": [
        "packages/*/src"
      ]
    },
```

File: packages/docs/package.json (hunk 16 above)
Comment (yogwoggf) [reply packages/docs/package.json]: This appears to be standard for any TS package. it's just that most will use `prepack` and only send JS files to npm, but since this is a local package, it builds for each install.

File: packages/docs/src/matchers/symmetric/index.ts (hunk 6 above)
Comment (Derpius) [reply packages/docs/src/matchers/symmetric/index.ts:34]: Looked into it, can't find a way

File: packages/site/src/components/object-renderer/object.tsx (hunk 14 above)
Comment (yogwoggf) [reply packages/site/src/components/object-renderer/object.tsx:5]: nit: dick

File: packages/docs/package.json (hunk 16 above)
Comment (Derpius) [reply packages/docs/package.json]: It was not required

File: packages/docs/src/matchers/symmetric/index.ts (hunk 7 above)
Comment (Derpius) [reply packages/docs/src/matchers/symmetric/index.ts]: Docusaurus handles this fine, both when building a release bundle and using hot reload

File: packages/docs/tsconfig.json (hunk 5 above)
Comment (Derpius) [reply packages/docs/tsconfig.json]: Composite was not required

**Comment by Derpius** [comment]
The main issue at the moment is that we've entirely lost hot reload in the site when editing the docs. I'm not sure how to resolve this at the moment but I'll open a ticket

**Review by Derpius — APPROVED** [review]
